File: 7_Semestre/LP/TTP_TP/Exemplos/Exemplo_Artigo/Streamlit/ST_Modelagem_Climatica_Python_Modules/modeling.py
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import numpy as np

class ClimateModel:
    def __init__(self, target_col='bio1', feature_cols=['bio12', 'bio4'], test_size=0.3, random_state=42):
        self.target_col = target_col
        self.feature_cols = feature_cols
        self.test_size = test_size
        self.random_state = random_state
        self.model = None
        self.metrics = None
        self.predictions = None
        self.test_data = None

    def train_model(self, df_clean: pd.DataFrame):
        """
        Treina o modelo de Regressão Linear Múltipla e calcula as métricas.
        """
        if df_clean is None or df_clean.empty:
            raise ValueError("DataFrame de entrada não pode ser nulo ou vazio.")

        # Preparação dos dados
        X = df_clean[self.feature_cols]
        y = df_clean[self.target_col]

        # Divisão em treino e teste
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=self.test_size, random_state=self.random_state
        )

        # Treinamento do modelo
        self.model = LinearRegression()
        self.model.fit(X_train, y_train)

        # Previsões
        self.predictions = self.model.predict(X_test)
        
        # Armazenar dados de teste para visualização
        self.test_data = {'X': X_test, 'y': y_test}

        # Cálculo das métricas
        r_squared = r2_score(y_test, self.predictions)
        rmse = np.sqrt(mean_squared_error(y_test, self.predictions))
        mae = mean_absolute_error(y_test, self.predictions)

        # Coeficientes
        coefficients = {
            'intercept': self.model.intercept_,
        }
        for feature, coef in zip(self.feature_cols, self.model.coef_):
            coefficients[feature] = coef

        self.metrics = {
            'r_squared': r_squared,
            'rmse': rmse,
            'mae': mae,
            'coefficients': coefficients
        }
        
        return self.model, self.metrics, self.predictions, self.test_data

    def get_metrics(self):
        return self.metrics

    def get_predictions(self):
        return self.predictions

    def get_test_data(self):
        return self.test_data

File: 7_Semestre/LP/TTP_TP/Exemplos/Exemplo_Artigo/R/R_Modelagem_Climactica_Modules/v2/modeling.r
# modeling.R

#' @title Funções de Modelagem Climática
#' @description Módulo responsável pela modelagem estatística dos dados climáticos

library(dplyr)
library(rsample)

#' Treinar modelo de regressão linear
#' @param df DataFrame com dados processados
#' @param target_var Variável alvo (padrão: bio1)
#' @param predictor_var Variável preditora (padrão: bio12)
#' @param prop Proporção para divisão treino/teste (padrão: 0.7)
#' @param seed Semente para reprodutibilidade (padrão: 42)
#' @return Lista com modelo, previsões, dados de teste e métricas
train_linear_regression <- function(df, target_var = "bio1", predictor_var = "bio12", prop = 0.7, seed = 42) {
  # Preparação dos dados
  data_model <- df %>% 
    dplyr::select(!!sym(target_var), !!sym(predictor_var)) %>% 
    na.omit()
  
  # Divisão dos dados
  set.seed(seed)
  split <- rsample::initial_split(data_model, prop = prop)
  train_data <- rsample::training(split)
  test_data <- rsample::testing(split)
  
  # Treinamento do Modelo de Regressão Linear
  formula <- as.formula(paste(target_var, "~", predictor_var))
  model <- lm(formula, data = train_data)
  
  # Previsões
  predictions <- predict(model, newdata = test_data)
  
  # Cálculo das métricas
  r_squared <- summary(model)$r.squared
  rmse <- sqrt(mean((test_data[[target_var]] - predictions)^2))
  
  # Cálculo do MAE (Mean Absolute Error)
  mae <- mean(abs(test_data[[target_var]] - predictions))
  
  # Coeficientes do modelo
  coefficients <- coef(model)
  
  metrics <- list(
    r_squared = r_squared,
    rmse = rmse,
    mae = mae,
    n_train = nrow(train_data),
    n_test = nrow(test_data),
    coefficients = coefficients
  )
  
  return(list(
    model = model,
    predictions = predictions,
    test_data = test_data,
    metrics = metrics
  ))
}

#' Gerar resumo do modelo formatado para UI
#' @param model_result Resultado do modelo de train_linear_regression
#' @return Lista com resumos formatados
get_model_summary <- function(model_result) {
  model <- model_result$model
  metrics <- model_result$metrics
  
  # Resumo estatístico
  model_summary <- summary(model)
  
  # Texto interpretativo
  interpretation <- paste(
    "O modelo explica", round(metrics$r_squared * 100, 2), 
    "% da variância na variável alvo.",
    "\n\nPara cada unidade de aumento na variável preditora,",
    "a variável alvo muda em", round(metrics$coefficients[2], 4), "unidades.",
    "\n\nO erro médio de previsão (RMSE) é de", round(metrics$rmse, 2), "unidades."
  )
  
  return(list(
    summary = model_summary,
    metrics = metrics,
    interpretation = interpretation
  ))
}

#' Calcular estatísticas descritivas
#' @param df DataFrame com dados
#' @param exclude_cols Colunas a excluir da análise
#' @return DataFrame com estatísticas
calculate_descriptive_stats <- function(df, exclude_cols = c("row", "col")) {
  # Selecionar apenas as colunas bioclimáticas
  bio_cols <- names(df)[!names(df) %in% exclude_cols]
  
  stats <- df %>%
    dplyr::select(dplyr::all_of(bio_cols)) %>%
    dplyr::summarise_all(list(
      Média = ~mean(., na.rm = TRUE),
      Mediana = ~median(., na.rm = TRUE),
      DP = ~sd(., na.rm = TRUE),
      Mín = ~min(., na.rm = TRUE),
      Máx = ~max(., na.rm = TRUE),
      Q1 = ~quantile(., 0.25, na.rm = TRUE),
      Q3 = ~quantile(., 0.75, na.rm = TRUE),
      NAs = ~sum(is.na(.))
    )) %>%
    t() %>%
    as.data.frame()
  
  colnames(stats) <- "Valor"
  stats$Variável <- rownames(stats)
  stats <- stats %>% dplyr::select(Variável, Valor)
  
  return(stats)
}

#' Calcular matriz de correlação
#' @param df DataFrame com dados
#' @param exclude_cols Colunas a excluir
#' @return Matriz de correlação
calculate_correlation_matrix <- function(df, exclude_cols = c("row", "col")) {
  bio_cols <- names(df)[!names(df) %in% exclude_cols]
  corr_matrix <- cor(df %>% dplyr::select(dplyr::all_of(bio_cols)), use = "complete.obs")
  return(corr_matrix)
}